Source/Evolve/WarpXEvolve.cpp

```diff
@@ -296,6 +296,7 @@ WarpX::Evolve (int numsteps)
         }
 
         if( do_electrostatic != ElectrostaticSolverAlgo::None ) {
+            if (warpx_py_beforeEsolve) warpx_py_beforeEsolve();
```

Comment:
Note that `warpx_py_beforeEsolve` is also called below. That call should presumably be moved into the `if` block for the EM solver.

Reply:
I think the current execution calls for both `warpx_py_beforeEsolve` and `warpx_py_afterEsolve` are only in `WarpX::OneStep_nosub`, which is called from line 180 in an `else if` block following `if (do_electrostatic != ElectrostaticSolverAlgo::None)`. So I think as is those callbacks are executed if and only if the EM solver is used.